tools/mlir-gen/MLIRGen.cpp

```diff
@@ -264,6 +309,54 @@ std::string MLIRGenerator::createMetadata() {
   return data;
 }
 
+void MLIRGenerator::computeFlops(ShapedType inputShape,
+                                 ShapedType outputShape) {
+  // Matmul flops = 2 * M * N * K = 2 * prod(inputDims) * N (outShape[1])
+  int64_t mkFlops = 1;
+  for (int i = 0, max = inputShape.getRank(); i < max; i++)
+    mkFlops *= inputShape.getDimSize(i);
+  int outRank = outputShape.getRank();
+  assert((outRank == 2 || outRank == 4) && "Invalid outRank");
+  // Tiled: N = NB * n = outShape[0] + outShape[3]
+  int64_t nFlops = outputShape.getDimSize(outRank - 1);
+  if (outRank > 2)
+    nFlops *= outputShape.getDimSize(1);
+  flops += 2 * mkFlops * nFlops;
+}
+
+Value MLIRGenerator::lowerNamedMatmul(Value input, Value weight, Value output) {
+  auto inputShape = cast<ShapedType>(input.getType());
+  auto weightShape = cast<ShapedType>(weight.getType());
+  auto outputShape = cast<ShapedType>(output.getType());
+  Value namedMatmul;
+  if (inputShape.getRank() == 2) {
+    namedMatmul = builder
+                      .create<linalg::MatmulOp>(
+                          loc, TypeRange{output.getType()},
+                          ValueRange{input, weight}, ValueRange{output})
+                      .getResult(0);
+  } else if (inputShape.getRank() == 4) {
```

Comment:
I can see long term value in going DAG route but I don't think we have necessary tools for that atm.
I'm not sure we should bother with named ops in these cases right now when we don't have proper ops to generate.

Also, this doesn't cover the VNNI case (mlir-gen opt `--vnni=2|4`) when rank is 5.

Reply:
One aspect of this work is to understand what we want and how we get `tpp-opt` to work with that. Doing both at the same time would be too much. So, if there's an intermediate state we can reach, good. If not, we may have to lower to a named ops format that we can't cope with it (if we know we can change the passes to do).

Reply:
> I can see long term value in going DAG route but I don't think we have necessary tools for that atm. I'm not sure we should bother with named ops in these cases right now when we don't have proper ops to generate.
> 
Another way we could lower this is by tiling + matmul
scf.for
    scf.for
       linalg.matmul ..
however, the subsequent passes need to handle it accordingly.

Reply:
> One aspect of this work is to understand what we want and how we get `tpp-opt` to work with that. Doing both at the same time would be too much. So, if there's an intermediate state we can reach, good. If not, we may have to lower to a named ops format that we can't cope with it (if we know we can change the passes to do).

Even we could have multiple ways of lowering controlled with option which will give more flexibility to play around.

Reply:
> Also, this doesn't cover the VNNI case (mlir-gen opt `--vnni=2|4`) when rank is 5.

Let's just make sure we cover the VNNI case.

Reply:
Okay maybe to capture my point a bit differently.

We want to switch to named ops to make things easier like matching, having clearer semantics, simplifying transforms and/or codegen. I don't see added value in anchoring around `mmt4d` which: is an arbitrary choice, needs more effort in matching compared to an equivalent generic, doesn't support VNNI cases anyway.

To me, this case shows we have some missing tooling around that space which might need to be addressed at Linalg level. But I'm not sure what it adds to `mlir-gen` as a tool.